tensorflow/core/kernels/diag_op.cc

```diff
@@ -160,39 +97,138 @@ class DiagPartOp : public OpKernel {
                    context->allocate_output(0, out_shape, &output));
+    functor::DiagPartFunctor<Device, T> diagPartFunc;
+    diagPartFunc(context->eigen_device<Device>(),
+                 out_shape.num_elements(),
+                 tensor.flat<T>().data(),
+                 output->flat<T>().data());
+    return;
+  }
+};
+
+// Implementation of the functor specialization for CPU.
+// 
+// According to the diagonal definition,
+// `output[i1,..., ik, i1,..., ik] = input[i1,..., ik]`,
+//
+// Let the rank of input is [s1,..., sk], then any offset of input's
+// pointer can be represent by coordinate [i1,..., ik],
+// where `index = i1*(s2*...*sk) + i2*(s3*...*sk) +... + ik`
+//
+// Let new_index is the offset of output's pointer with coordinate 
+// [i1,..., ik, i1,..., ik], then we have
+// `new_index = i1*(s2*...sk*s1*...*sk) + i2*(s3*...*sk*s1*...*sk) +... + \
+//              ik*(s1*...*sk) + i1*(s2*...*sk) + i2*(s3*...*sk) +... + ik
+//            = (i1*(s2*...*sk) + i2*(s3*...*sk) +... + ik) * (1 + s1*...*sk)
+//            = index * (1 + s1*...*sk)
+//
+// Let `size = s1*...*sk`, we finally have `new_index = index * (1 + size)`,
+// which is the transfer function we use below.
+// This trick make our implementations clear and easy to be parallel.
+namespace functor {
+template <typename T>
+struct DiagFunctor<CPUDevice, T> {
+  void operator() (const CPUDevice& device, const int64 size,
+                   const T* in, T* out) {
+    std::fill(out, out + size * size, T());
+    for (int64 index = 0; index < size; index++) {
+      out[(1 + size) * index] = in[index];
+    }
+  }
+};
 
-    switch (num_dims) {
-      case 2:
-        output->tensor<T, 1>() = output->tensor<T, 1>().generate(
-          DiagonalExtractor<T, 1>(tensor));
-        break; 
-      case 4:
-        output->tensor<T, 2>() = output->tensor<T, 2>().generate(
-          DiagonalExtractor<T, 2>(tensor));
-        break;
-      case 6:
-        output->tensor<T, 3>() = output->tensor<T, 3>().generate(
-          DiagonalExtractor<T, 3>(tensor));
-        break;      
-      default:
-        context->SetStatus(errors::Unimplemented(
-          "Diagonal of rank ", num_dims, " tensor is not supported yet."));
-        return;
+template <typename T>
+struct DiagPartFunctor<CPUDevice, T> {
+  void operator() (const CPUDevice& device, const int64 size,
+                   const T* in, T* out) {
+    for (int64 index = 0; index < size; index++) {
+      out[index] = in[(1 + size) * index];
     }
   }
 };
+}  // namespace functor
+
 
-#define REGISTER_DIAGPARTOP(T) \
-  REGISTER_KERNEL_BUILDER( \
-      Name("DiagPart").Device(DEVICE_CPU).TypeConstraint<T>("T"), DiagPartOp<T>)
+// Register the CPU kernels.
+#define REGISTER_DIAGOP(T)                                    \
+  REGISTER_KERNEL_BUILDER(                                    \
+      Name("Diag").Device(DEVICE_CPU).TypeConstraint<T>("T"), \
+      DiagOp<CPUDevice, T>)
+
+REGISTER_DIAGOP(double);
```

Comment:
could you change these to use the macros from core/framework/register_types.h  (TF_CALL_double, TF_CALL_float, etc.)?  Same for the REGISTER_DIAGPARTOP?